javaagent-extension-api/src/test/java/io/opentelemetry/javaagent/bootstrap/internal/StructuredConfigPropertiesBridgeTest.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.bootstrap.internal;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import io.opentelemetry.sdk.autoconfigure.spi.ConfigProperties;
+import io.opentelemetry.sdk.autoconfigure.spi.internal.StructuredConfigProperties;
+import io.opentelemetry.sdk.extension.incubator.fileconfig.FileConfiguration;
+import java.io.ByteArrayInputStream;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import org.junit.jupiter.api.Test;
+
+class StructuredConfigPropertiesBridgeTest {
```

Comment:
StructuredConfigPropertiesBridge is my proposal for how we keep a sane configuration story in the agent / instrumentation despite the SDK having either flat sys prop / env var or structured YAML configuration options.

The idea is the implement the `ConfigProperties` interface using a backing `StructuredConfigProperties` instance, and have special logic to determine how the [standard java agent](https://opentelemetry.io/docs/zero-code/java/agent/disable/ properties map to a structured representation. 
- Only properties starting with `otel.instrumentation.` resolve
- `otel.instrumentation.` refers to the node at [`.instrumentation.java`](https://github.com/open-telemetry/opentelemetry-configuration/blob/16f8f9943df509fc7e9ed3cb4e6e358d0bf8058f/examples/kitchen-sink.yaml#L433-L436) in the declarative config data model
- To resolve a property, we take the portion after `otel.instrumentation` and break it up into segments. For each segment (N-1) we walk down the StructuredConfigProperties tree, and read the final segment from the resolve node.

So given the following YAML, reading `configProperties.getBoolean("otel.instrumentation.common.default-enabled")` resolves true:
```
instrumentation:
  java:
    common:
      default-enabled: true
```

Reply:
See https://github.com/open-telemetry/opentelemetry-java-examples/pull/492 for a demonstration

(Note, requires building the javaagent locally, including this [PR](https://github.com/open-telemetry/opentelemetry-java/pull/6733) from `opentelemetry-java`)

Reply:
long-term, would we bridge in the other direction? i.e. replace ConfigProperties usage with StructuredConfigProperties everywhere in the Java agent, and have a bridge from ConfigProperties -> StructuredConfigProperties?

Reply:
That's certainly an option. I think we ought to talk through options and repercussions as a group.

One thing to note is that the access patterns for `StructuredConfigProperties` and `ConfigProperties` are quite different.

For example, accessing `otel.instrumentation.common.default-enabled` is something like:
- ConfigProperties: `config.getBoolean("otel.instrumentation.common.default-enabled")`
- StructuredConfigProperties (assuming you're starting at the `.instrumentation.java` node) `config.getStructured("config").getBoolean("default-enabled")` 
  - needs additional null handling each time you walk down a node
  - diverges further for each level of nesting

Also need to remember that there's now a specification for an instrumentation config API (and java PR here https://github.com/open-telemetry/opentelemetry-java/pull/6549). The java agent works around the historic lack of availability of such a thing, and we'll need to think about how / when / (if?) to cut over. There's lots of instrumentation that reads config - is it feasible for a one time cutover to the new APIs, or do we need to plan for some sort of iterative approach?

Reply:
> needs additional null handling each time you walk down a node

what do you think of optional return types that support subsequent accessors without null check?

> diverges further for each level of nesting

I didn't follow what you mean?

> we'll need to think about how / when / (if?) to cut over. There's lots of instrumentation that reads config - is it feasible for a one time cutover to the new APIs, or do we need to plan for some sort of iterative approach?

if we do some initial prototyping and like it, we could do a one-time cutover in the 3.0.0 release

Reply:
>> diverges further for each level of nesting
> I didn't follow what you mean?

Just that the difference in syntax grows the more levels of nesting you have. 

> what do you think of optional return types that support subsequent accessors without null check?

Well we need some sort of syntactic sugar to not it make so painful, that's for sure.

The problem with returning `Optional` is that it creates a tri-state (optional-present, optional-empty, and null). While we can trust that the core implementation of `StructuredConfigProperties` never returns null, other user-provided interfaces might not be so well behaved. 

Could also define a "get or default" overload of the form: `getStructured(String name, StructuredConfigProperties defaultProperties)`

If we could this with a `EMPTY_CONFIG` constant, walking the tree becomes:

```
config.getStructured("common", EMPTY_CONFIG).getBoolean("default-enabled", true);
```

Not bad.

> if we do some initial prototyping and like it, we could do a one-time cutover in the 3.0.0 release

That's true, but I doubt that the instrumentation config API will be stable in time for 3.0.0 which I assume is happening at the end of the year. So then we'll need to consider whether we're comfortable depending on an experimental API, or should wait till 4.0.0 for a cutover.

Reply:
I see these as feasible variants

1. Optional:

```java
config.getStructured("common")
  .flatMap(c -> c.getStructured("foo")
  .flatMap(c -> c.getBoolean("default-enabled", true));
```

2. getStructured never returns null (but maybe has an `isPresent` method) - I like that one

```java
config.getStructured("common").getStructured("foo").getBoolean("default-enabled", true);
```

It's true that custom implementations may still return `null` (for any of the above) - but they could also throw exceptions.

Reply:
> 1. Optional:

The SDK has no precedent for returning optional instead of null. I believe that was an intentional design decision long ago, and I'm not interested in breaking symmetry.

> 2. getStructured never returns null (but maybe has an isPresent method) - I like that one

Users need a way to tell determine the when a key was present and empty vs. null, so never returning null (or some equivalent) is not an option. We could add isPresent, but that's the same as calling `getString("foo") != null` so no improvement in user calling ergonomics.

PR here to add getStructured default method, which I think has good ergonomics and is symmetric with existing patterns in opentelemetry-java: https://github.com/open-telemetry/opentelemetry-java/pull/6759

Reply:
So it would look like this for a [real example](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/798bdd5569e282b385100270aa25c2976071070d/instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/config/internal/CommonConfig.java#L44) - looks too heavy IMO:

```java
structuredConfigProps
                .getStructured("otel", empty())
                .getStructured("instrumentation", empty())
                .getStructured("http", empty())
                .getStructured("client", empty())
                .getMap("capture-request-headers"))
```

Reply:
There's abstractions in place to read common properties so that only a few places in code have to do deep tree walking.